docs/basics.txt

```diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+     defaultPassword: password,
+     credentialsDb: conf/janusgraph-credentials-server.properties
+    }
+ }
+```
+
+Verify that basic authentication is configured correctly. For example
+
+```
+curl -v -XPOST http://localhost:8182 -d '{"gremlin": "g.V().count()"}'
+```
+
+should return a 401 if the authentication is configured correctly and
+
+```
+curl -v -XPOST http://localhost:8182 -d '{"gremlin": "g.V().count()"}' -u user:password
+```
+should return a 200 and the result of 4 if authentication is configured correctly.
+
+===== HTTP HMAC authentication
+
+Another authentication method avaiablable, designed to amortize the number of rounds required to encrypt the password in basic auth, is HMAC authentication provided by JanusGraph. This is a token based authentication designed to be used over HTTP. You first acquire a token via the `/session` endpoint and then use that to authenticate.
+
+To enable HMAC token authentiation, the gremlin-server.yaml configuration should include the following properties
+
+```
+ authentication: {
+   authenticator: org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.HMACAuthenticator,
+   authenticationHandler: org.janusgraph.graphdb.tinkerpop.gremlin.server.handler.HttpHMACAuthenticationHandler,
+   config: {
+     defaultUsername: user,
+     defaultPassword: password,
+     hmacSecret: secret,
+     credentialsDb: conf/janusgraph-credentials-server.properties
+   }
+ }
+```
+IMPORTANT: Note the hmacSecret here. This should be the same across all running JanusGraph servers if you want to be able to use the same HMAC token on each server.
+
+This creates a `/session` endpoint that provides a token that expires after an hour by default. This timeout for the token can be configured through the `tokenTimeout` configuration option in the `authentication.config` map. This value is a Long value and in milliseconds.
+
+You can obtain the token using curl by issuing a get request to the `/session` endpoint. For example
+
+```
+curl http://localhost:8182/session -XGET -u user:password
+
+{"token": "dXNlcjoxNTA5NTQ2NjI0NDUzOkhrclhYaGhRVG9KTnVSRXJ5U2VpdndhalJRcVBtWEpSMzh5WldqRTM4MW89"}
+```
+
+You can then use that token for authentication by using the "Authorization: Token" header. For example
+
+```
+curl -v http://localhost:8182/session -XPOST -d '{"gremlin": "1+1"}' -H "Authorization: Token dXNlcjoxNTA5NTQ2NjI0NDUzOkhrclhYaGhRVG9KTnVSRXJ5U2VpdndhalJRcVBtWEpSMzh5WldqRTM4MW89"
+```
+
+==== Authentication over WebSocket
+Authentication over WebSocket occurs through a Simple Authentication and Security Layer (https://en.wikipedia.org/wiki/Simple_Authentication_and_Security_Layer[SASL]) mechanism.
+
+
+To enable SASL authentication include the following configuration in the gremlin-server.yaml
+
+```
+authentication: {
+  authenticator: org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.JanusGraphSimpleAuthenticator,
+  authenticationHandler: org.apache.tinkerpop.gremlin.server.handler.HttpBasicAuthenticationHandler,
```

Comment:
I'm getting an error with this config when trying to connect in Gremlin console. Should it be the SASL authenticator[Handler] here?

```
io.netty.channel.ChannelPipelineException: org.apache.tinkerpop.gremlin.server.handler.HttpBasicAuthenticationHandler is not a @Sharable handler, so can't be added or removed multiple times.
```